File: frontend/src/types/events.ts
// Real-time retail event types — mirrors backend schemas/events.py exactly

export type EventType =
  | 'sale.recorded'
  | 'sale.spike_detected'
  | 'sale.drop_detected'
  | 'inventory.stock_low'
  | 'inventory.stock_critical'
  | 'inventory.stock_out'
  | 'inventory.reorder_triggered'
  | 'inventory.reorder_approved'
  | 'demand.forecast_updated'
  | 'demand.anomaly'
  | 'demand.trend_change'
  | 'business.revenue_milestone'
  | 'business.slow_mover_alert'
  | 'business.fast_mover_alert'
  | 'business.price_anomaly'
  | 'customer.activity'
  | 'customer.order_placed'
  | 'agent.decision'
  | 'system.agent_heartbeat'
  | 'system.model_retrained'

export type EventSeverity = 'info' | 'warning' | 'critical' | 'success'
export type AgentRole = 'demand' | 'intelligence' | 'manager'
export type DecisionPriority = 'critical' | 'high' | 'medium' | 'low'

export interface RetailEvent {
  event_id: string
  event_type: EventType
  shop_id: string
  product_id: string | null
  product_name: string | null
  severity: EventSeverity
  title: string
  message: string
  payload: Record<string, unknown>
  action_required: boolean
  action_label: string | null
  action_endpoint: string | null
  timestamp: string
  acknowledged: boolean
}

export interface AlertNotification {
  alert_id: string
  shop_id: string
  severity: EventSeverity
  title: string
  body: string
  product_id: string | null
  product_name: string | null
  recommended_action: string
  business_impact: string
  action_label: string | null
  action_endpoint: string | null
  created_at: string
  acknowledged: boolean
  source_event_ids: string[]
}

export interface AgentDecisionPayload {
  decision_id: string
  agent: AgentRole
  priority: DecisionPriority
  action: string
  action_label: string | null
  action_endpoint: string | null
  confidence: number
  reasoning: string
  [key: string]: unknown
}

export interface LiveKPI {
  shop_id: string
  as_of: string
  revenue_today: number
  orders_today: number
  units_today: number
  unacked_alerts: number
  agent_decisions_recent: number
  critical_events_recent: number
  agents_active: boolean
  event_stream_subscribers: number
}

File: backend/app/schemas/events.py
"""
Retail Event Schema — typed events that flow through the event bus.
Every retail action in the store generates one of these events.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class EventType(str, Enum):
    # Sales events
    SALE_RECORDED = "sale.recorded"
    SALE_SPIKE_DETECTED = "sale.spike_detected"
    SALE_DROP_DETECTED = "sale.drop_detected"

    # Inventory events
    STOCK_LOW = "inventory.stock_low"
    STOCK_CRITICAL = "inventory.stock_critical"
    STOCK_OUT = "inventory.stock_out"
    REORDER_TRIGGERED = "inventory.reorder_triggered"
    REORDER_APPROVED = "inventory.reorder_approved"

    # Demand events
    DEMAND_FORECAST_UPDATED = "demand.forecast_updated"
    DEMAND_ANOMALY = "demand.anomaly"
    DEMAND_TREND_CHANGE = "demand.trend_change"

    # Business events
    REVENUE_MILESTONE = "business.revenue_milestone"
    SLOW_MOVER_ALERT = "business.slow_mover_alert"
    FAST_MOVER_ALERT = "business.fast_mover_alert"
    PRICE_ANOMALY = "business.price_anomaly"

    # Customer events
    CUSTOMER_ACTIVITY = "customer.activity"
    ORDER_PLACED = "customer.order_placed"

    # Agent events
    AGENT_DECISION = "agent.decision"
    AGENT_HEARTBEAT = "system.agent_heartbeat"
    MODEL_RETRAINED = "system.model_retrained"


class EventSeverity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"
    SUCCESS = "success"


class RetailEvent(BaseModel):
    event_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    event_type: EventType
    shop_id: str
    product_id: Optional[str] = None
    product_name: Optional[str] = None
    severity: EventSeverity = EventSeverity.INFO
    title: str
    message: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    action_required: bool = False
    action_label: Optional[str] = None
    action_endpoint: Optional[str] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    acknowledged: bool = False

    def to_sse(self) -> str:
        """Format as Server-Sent Event string."""
        import json
        data = self.model_dump(mode="json")
        return f"data: {json.dumps(data)}\n\n"


class AlertNotification(BaseModel):
    """Proactive AI-generated alert surfaced to the store owner."""
    alert_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:10])
    shop_id: str
    severity: EventSeverity
    title: str
    body: str
    product_id: Optional[str] = None
    product_name: Optional[str] = None
    recommended_action: str
    business_impact: str
    action_label: Optional[str] = None
    action_endpoint: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    acknowledged: bool = False
    source_event_ids: List[str] = Field(default_factory=list)


class AgentRole(str, Enum):
    DEMAND = "demand"
    INTELLIGENCE = "intelligence"
    MANAGER = "manager"


class DecisionPriority(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class AgentDecision(BaseModel):
    """Structured decision produced by an AI agent with full reasoning chain."""
    decision_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:10])
    shop_id: str
    agent: AgentRole
    priority: DecisionPriority
    title: str
    reasoning: str
    action: str
    action_label: Optional[str] = None
    action_endpoint: Optional[str] = None
    confidence: float = Field(default=0.8, ge=0.0, le=1.0)
    product_id: Optional[str] = None
    product_name: Optional[str] = None
    source_event_ids: List[str] = Field(default_factory=list)
    payload: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    executed: bool = False
